bridge/github/export.go

```diff
@@ -0,0 +1,654 @@
+	cachedLabels map[string]string
+}
+
+// Init .
+func (ge *githubExporter) Init(conf core.Configuration) error {
+	//TODO: initialize with multiple tokens
+	ge.conf = conf
+	ge.identityToken = make(map[string]string)
+	ge.identityClient = make(map[string]*githubv4.Client)
+	ge.cachedIDs = make(map[string]string)
+	ge.cachedLabels = make(map[string]string)
+	return nil
+}
+
+func (ge *githubExporter) getIdentityClient(id string) (*githubv4.Client, error) {
+	client, ok := ge.identityClient[id]
+	if ok {
+		return client, nil
+	}
+
+	token, ok := ge.identityToken[id]
+	if !ok {
+		return nil, fmt.Errorf("unknown identity token")
+	}
+
+	return buildClient(token), nil
+}
+
+// ExportAll export all event made by the current user to Github
+func (ge *githubExporter) ExportAll(repo *cache.RepoCache, since time.Time) error {
+	user, err := repo.GetUserIdentity()
+	if err != nil {
+		return err
+	}
+
+	ge.identityToken[user.Id()] = ge.conf[keyToken]
+
+	// get repository node id
+	ge.repositoryID, err = getRepositoryNodeID(
+		ge.conf[keyOwner],
+		ge.conf[keyProject],
+		ge.conf[keyToken],
+	)
+
+	if err != nil {
+		return err
+	}
+
+	allBugsIds := repo.AllBugsIds()
+bugLoop:
+	for _, id := range allBugsIds {
+		b, err := repo.ResolveBug(id)
+		if err != nil {
+			return err
+		}
+
+		snapshot := b.Snapshot()
+
+		// ignore issues created before since date
+		if snapshot.CreatedAt.Before(since) {
+			continue
+		}
+
+		// if identity participated in a bug
+		for _, p := range snapshot.Participants {
+			for userId := range ge.identityToken {
+				if p.Id() == userId {
+					// try to export the bug and it associated events
+					if err := ge.exportBug(b, since); err != nil {
+						return err
+					}
+
+					continue bugLoop
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// exportBug publish bugs and related events
+func (ge *githubExporter) exportBug(b *cache.BugCache, since time.Time) error {
+	snapshot := b.Snapshot()
+
+	var bugGithubID string
+	var bugGithubURL string
+	var bugCreationHash string
+
+	// Special case:
+	// if a user try to export a bug that is not already exported to Github (or imported
+	// from Github) and he is not the author of the bug. There is nothing we can do.
+
+	// first operation is always createOp
+	createOp := snapshot.Operations[0].(*bug.CreateOperation)
+	author := createOp.GetAuthor()
+
+	// get github bug ID
+	githubID, ok := createOp.GetMetadata(keyGithubId)
+	if ok {
+		githubURL, ok := createOp.GetMetadata(keyGithubId)
+		if !ok {
+			// if we find github ID, github URL must be found too
+			panic("expected to find github issue URL")
+		}
+
+		// will be used to mark operation related to a bug as exported
+		bugGithubID = githubID
+		bugGithubURL = githubURL
+
+	} else {
+		// check that we have a token for operation author
+		client, err := ge.getIdentityClient(author.Id())
+		if err != nil {
+			// if bug is still not exported and user cannot author bug stop the execution
+
+			// TODO: maybe print a warning ?
```

Comment:
I'd say print a warning for now.

If at some point we consider that the exporter is a low level function that multiple UI can use, we will have it return a channel where this kind of event are pushed (see https://github.com/MichaelMure/git-bug/blob/master/bug/bug_actions.go#L53 for example)

Reply:
done !